File: automation-test-playwright/pages/admin/AdminSaleContractListPage.ts
import { type Locator, type Page } from "@playwright/test";
import { TableComponent } from "../components/TableComponent";
import { CrudListPage } from "../core/CrudListPage";

export class AdminSaleContractListPage extends CrudListPage {
  protected readonly path = "/admin/sale-contract/list";
  readonly addButton: Locator;
  readonly tableBody: Locator;
  private readonly table: TableComponent;

  /**
   * Initializes this page object.
   */
  constructor(page: Page) {
    super(page);
    this.addButton = this.page.locator(".btn-add");
    this.tableBody = this.page.locator("#saleContractTableBody");
    this.table = new TableComponent(page, "#saleContractTableBody");
  }

  async waitForLoaded(): Promise<void> {
    await this.page.waitForURL(/\/admin\/sale-contract\/(list|search)/);
    await this.waitForVisible(this.tableBody);
  }

  async waitForTableData(): Promise<void> {
    await this.table.waitForDataOrEmpty();
  }

  async openAddForm(): Promise<void> {
    await this.addButton.click();
  }

  /**
   * Applies the customer filter.
   */
  async filterByCustomer(customerId: number | string): Promise<void> {
    await this.selectFilter("customerId", String(customerId));
  }

  /**
   * Applies the building filter.
   */
  async filterByBuilding(buildingId: number | string): Promise<void> {
    await this.selectFilter("buildingId", String(buildingId));
  }

  /**
   * Applies the staff filter.
   */
  async filterByStaff(staffId: number | string): Promise<void> {
    await this.selectFilter("staffId", String(staffId));
  }

  /**
   * Applies the status filter.
   */
  async filterByStatus(status: "0" | "1"): Promise<void> {
    await this.selectFilter("status", status);
  }

  /**
   * Submits the active filters.
   */
  async submitFilters(): Promise<void> {
    await this.search();
  }

  rowBySaleContractText(text: string): Locator {
    return this.table.rowByText(text);
  }

  async openDetail(text: string): Promise<void> {
    await this.clickRowLink(text, "/admin/sale-contract/");
  }

  async openEdit(text: string): Promise<void> {
    await this.clickRowLink(text, "/admin/sale-contract/edit/");
  }

  /**
   * Deletes sale contract through the UI.
   */
  async deleteSaleContract(text: string): Promise<void> {
    await this.actionButton(this.rowBySaleContractText(text), "delete").click();
  }

  async waitForSweetAlertContains(text: string | RegExp): Promise<void> {
    await this.waitForSweetAlertContainsText(text);
  }
}

File: automation-test-selenium/src/main/java/org/suntower/pages/admin/AdminSaleContractListPage.java
package org.suntower.pages.admin;

import org.openqa.selenium.By;
import org.openqa.selenium.JavascriptExecutor;
import org.openqa.selenium.WebDriver;
import org.openqa.selenium.WebElement;
import org.openqa.selenium.support.ui.ExpectedConditions;
import org.suntower.pages.components.TableComponent;
import org.suntower.pages.core.CrudListPage;

public class AdminSaleContractListPage extends CrudListPage {
  private final By tableBody = css("#saleContractTableBody");
  private final TableComponent table;

  public AdminSaleContractListPage(WebDriver driver) {
    super(driver);
    this.path = "/admin/sale-contract/list";
    this.table = new TableComponent(driver, "#saleContractTableBody");
  }

  public void waitForLoaded() {
    waitForUrlMatches(".*/admin/sale-contract/(list|search).*");
    waitForVisible(tableBody);
  }

  public void waitForTableData() {
    table.waitForDataOrEmpty();
  }

  public WebElement rowBySaleContractText(String text) {
    return table.rowByText(text);
  }

  public void openDetail(String text) {
    String href = rowBySaleContractText(text).findElement(By.cssSelector("a[href*='/admin/sale-contract/']")).getAttribute("href");
    visit(href);
  }

  public void filterByCustomer(long customerId) {
    setFormControlValue("customerId", String.valueOf(customerId));
  }

  public void filterByBuilding(long buildingId) {
    setFormControlValue("buildingId", String.valueOf(buildingId));
  }

  @Override
  public void search() {
    wait.until(ExpectedConditions.presenceOfElementLocated(css("form")));
    ((JavascriptExecutor) driver).executeScript("document.querySelector('form').requestSubmit();");
    waitForTableData();
  }

  private void setFormControlValue(String fieldName, String value) {
    wait.until(ExpectedConditions.presenceOfElementLocated(By.name(fieldName)));
    ((JavascriptExecutor) driver)
        .executeScript(
            "const el = document.querySelector('[name=\"' + arguments[0] + '\"]');"
                + "el.value = arguments[1];"
                + "el.dispatchEvent(new Event('input', { bubbles: true }));"
                + "el.dispatchEvent(new Event('change', { bubbles: true }));",
            fieldName,
            value);
  }
}
